File: src/test/java/com/github/njuro/jboard/post/decorators/GreentextDecoratorTest.java
package com.github.njuro.jboard.post.decorators;

import static com.github.njuro.jboard.common.Constants.GREENTEXT_END;
import static com.github.njuro.jboard.common.Constants.GREENTEXT_START;
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

class GreentextDecoratorTest extends PostDecoratorTest {

  @Override
  protected PostDecorator initDecorator() {
    return new GreentextDecorator();
  }

  @ParameterizedTest
  @ValueSource(
      strings = {
        ">text",
        ">  text",
        "    >longer text > another",
        "  > some longer text",
        ">>>multiple quotes",
        ">multiline\r\n>text"
      })
  void testValidGreentext(String input) {
    decoratePost(input);
    assertThat(PostDecoratorTest.post.getBody())
        .containsSubsequence(GREENTEXT_START, GREENTEXT_END);
  }

  @ParameterizedTest
  @ValueSource(strings = {"aaa>text", "    aaaa > text", " aaa>text "})
  void testInvalidGreentext(String input) {
    decoratePost(input);
    assertThat(PostDecoratorTest.post.getBody()).isEqualTo(input);
  }
}

File: src/test/kotlin/com/github/njuro/jard/post/decorators/GreentextDecoratorTest.kt
package com.github.njuro.jard.post.decorators

import com.github.njuro.jard.common.Constants
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContainInOrder
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

internal class GreentextDecoratorTest : PostDecoratorTest() {
    override fun initDecorator() = GreentextDecorator()

    @ParameterizedTest
    @ValueSource(
        strings = [
            ">text",
            ">  text",
            "    >longer text > another",
            "  > some longer text",
            ">>>multiple quotes",
            ">multiline\r\n>text"
        ]
    )
    fun `valid greentext`(input: String) {
        decorate(input).shouldContainInOrder(Constants.GREENTEXT_START, Constants.GREENTEXT_END)
    }

    @ParameterizedTest
    @ValueSource(
        strings = [
            "aaa>text",
            "    aaaa > text",
            " aaa>text "
        ]
    )
    fun `invalid greentext`(input: String) {
        decorate(input) shouldBe input
    }
}